pkg/awsutils/awsutils.go

```diff
@@ -343,19 +358,61 @@ func (cache *EC2InstanceMetadataCache) initWithEC2Metadata() error {
 	}
 	log.Debugf("Found vpc-ipv4-cidr-block: %s ", cache.vpcIPv4CIDR)
 
-	// retrieve vpc-ipv4-cidr-blocks
+	// initSGIDs retrieves security groups
+	err = cache.initSGIDs(mac)
+	if err != nil {
+		return err
+	}
+
+	// initVPCIPv4CIDRs retrieves VPC IPv4 CIDR blocks
+	err = cache.initVPCIPv4CIDRs(mac)
+	if err != nil {
+		return err
+	}
+
+	// refresh security groups and VPC CIDR blocks in the background
+	go wait.Forever(func() { cache.initSGIDs(mac) }, 30*time.Second)
+	go wait.Forever(func() { cache.initVPCIPv4CIDRs(mac) }, 30*time.Second)
+
+	select {
+	case <-ctx.Done():
+		return nil
+	}
+}
+
+// initSGIDs retrieves security groups
+func (cache *EC2InstanceMetadataCache) initSGIDs(mac string) error {
+	metadataSGIDs, err := cache.ec2Metadata.GetMetadata(metadataMACPath + mac + metadataSGs)
+	if err != nil {
+		awsAPIErrInc("GetMetadata", err)
+		log.Errorf("Failed to retrieve security-group-ids data from instance metadata service")
+		return errors.Wrap(err, "get instance metadata: failed to retrieve security-group-ids")
+	}
+
+	sgIDs := strings.Fields(metadataSGIDs)
+	cache.securityGroups.reset()
+	for _, sgID := range sgIDs {
+		log.Debugf("Found security-group id: %s", sgID)
+		cache.securityGroups.add(sgID)
+	}
+	return nil
+}
+
+// initVPCIPv4CIDRs retrieves VPC IPv4 CIDR blocks
+func (cache *EC2InstanceMetadataCache) initVPCIPv4CIDRs(mac string) error {
 	metadataVPCIPv4CIDRs, err := cache.ec2Metadata.GetMetadata(metadataMACPath + mac + metadataVPCcidrs)
+
 	if err != nil {
 		awsAPIErrInc("GetMetadata", err)
 		log.Errorf("Failed to retrieve vpc-ipv4-cidr-blocks from instance metadata service")
-		return errors.Wrap(err, "get instance metadata: failed to retrieve vpc-ipv4-cidr-block data")
+		return errors.Wrap(err, "get instance metadata: failed to retrieve vpc-ipv4-cidr-blocks data")
 	}
 
 	vpcIPv4CIDRs := strings.Fields(metadataVPCIPv4CIDRs)
-
+	cache.vpcIPv4CIDRs.reset()
 	for _, vpcCIDR := range vpcIPv4CIDRs {
 		log.Debugf("Found VPC CIDR: %s", vpcCIDR)
```

Comment:
The above is now going to be output every 30 seconds, one for each subnet associated with the node. Instead of just calling `reset()` on the StringSet above, you could consider determining what the set difference is between the values you fetched from IMDS and then only outputting a log message (or changing the value of the StringSet) if something changed.